File: server/ops/environment/source.js
// Heartbeat source: "is system load unusually high or climbing without a
// clear cause" (root CLAUDE.md's Operational Awareness item 5). Leaf
// module — only baseline.js (a leaf) plus schedule-store.js for its own
// prior checkState, same shape as heartbeat/sources/jobs-source.js.
//
// **Own-checkState dedup, same discipline jobs-source.js already
// established and heartbeat/CLAUDE.md documents at length — required here
// for the identical reason.** A sustained-high-load condition can easily
// hold true across many ticks; without remembering "I already reported
// THIS ongoing situation," every tick would produce a fresh finding, a
// fresh spent decision.js model call, and a fresh notification for the
// exact same still-true condition. `check()` only returns a Finding when
// the reported KIND changes (nothing -> cpu, cpu -> memory, memory ->
// nothing, ...) — a genuinely new situation, never the same one repeated.

import { checkForAnomaly } from './baseline.js';
import { getItem as getScheduleItem } from '../../heartbeat/schedule-store.js';

export const SOURCE_ID = 'environment';
const CHECK_INTERVAL_MS = 5 * 60 * 1000; // cheap SQLite read only — no model call itself

/** One fixed item — system load isn't per-entity the way a job or a memory commitment is. */
export async function listItems() {
  return [{ itemKey: 'system-load', intervalMs: CHECK_INTERVAL_MS }];
}

export async function check(itemKey) {
  const anomaly = checkForAnomaly();
  const kind = anomaly ? JSON.parse(anomaly.detail).kind : null;

  const prior = getScheduleItem(SOURCE_ID, itemKey)?.checkState;
  if (!anomaly) {
    if (prior?.reportedKind) return { finding: null, checkState: null }; // it settled — clear the memory of the past report
    return { finding: null };
  }

  if (prior?.reportedKind === kind) {
    return { finding: null }; // the same still-ongoing condition — already reported
  }

  return {
    finding: { summary: anomaly.summary, detail: anomaly.detail },
    checkState: { reportedKind: kind, reportedAt: new Date().toISOString() },
  };
}

export const source = { id: SOURCE_ID, defaultIntervalMs: CHECK_INTERVAL_MS, listItems, check };

File: backend/jarvis/ops/environment/source.py
"""System load, as something worth noticing.

One item: load is not per-entity the way a job or a commitment is.

Its own `checkState` dedup, for the reason the source contract spells out — a
sustained-high-load condition holds true for as long as it lasts, and a finding
per tick would mean a model call and a notification every few minutes for one
situation the user already knows about. A finding is produced only when the KIND
of anomaly changes (nothing to CPU, CPU to memory, back to nothing).
"""

from __future__ import annotations

import json
from typing import Any

from ...heartbeat.sources.registry import Source
from . import baseline

SOURCE_ID = "environment"
CHECK_INTERVAL_MS = 5 * 60 * 1000


def list_items() -> list[dict[str, Any]]:
    return [{"itemKey": "system-load", "intervalMs": CHECK_INTERVAL_MS}]


def check(item_key: str) -> dict[str, Any]:
    from ...heartbeat import schedule_store

    anomaly = baseline.check_for_anomaly()
    kind = json.loads(anomaly["detail"])["kind"] if anomaly else None

    item = schedule_store.get_item(SOURCE_ID, item_key)
    reported = ((item or {}).get("checkState") or {}).get("reportedKind")

    if anomaly is None:
        # Settled: forget the past report, so a genuine recurrence is reported
        # again rather than suppressed forever by a stale memory.
        return {"finding": None, "checkState": None} if reported else {"finding": None}

    if reported == kind:
        return {"finding": None}         # the same situation, still true

    return {"finding": {"summary": anomaly["summary"], "detail": anomaly["detail"]},
            "checkState": {"reportedKind": kind}}


source = Source(id=SOURCE_ID, default_interval_ms=CHECK_INTERVAL_MS,
                list_items=list_items, check=check)
